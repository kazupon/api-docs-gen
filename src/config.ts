import type { Resolver, Processor } from './types'

/**
 * Configration
 */
export interface Config {
  /**
   * markdown link reference resolver
   */
  resolver?: Resolver
  /**
   * processor
   */
  processor: Processor
}
