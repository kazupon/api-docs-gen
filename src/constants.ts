/**
 * Constant of `--genereate-style` options
 */
export const GENERATE_STYLES = {
  prefix: 'prefix',
  directory: 'directory'
} as const

/**
 * Type of `--generate-style` options
 *
 * @remarks See {@link GENERATE_STYLES} contents
 */
export type GenerateStyle = typeof GENERATE_STYLES[keyof typeof GENERATE_STYLES]
