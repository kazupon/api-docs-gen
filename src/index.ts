export { getSafePathFromDisplayName, escapeText } from './utils'
export * from './builder'
export * from './config'
export {
  multi as multiProcessor,
  toc as tocProcessor,
  getCustomTags,
  getDocSectionContent
} from './processor'
export { multi as multiResolver, toc as tocResolver } from './resolver'
export { generate } from './generator'
