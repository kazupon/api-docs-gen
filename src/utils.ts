export const isString = (val: unknown): val is string => typeof val === 'string'
