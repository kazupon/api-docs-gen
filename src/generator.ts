import path from 'path'
import { ApiModel } from '@microsoft/api-extractor-model'
import { debug as Debug } from 'debug'
import { loadPackage } from './resolver'
import { isString, mkdir, writeFile } from './utils'

import type { Config, GenerateStyle } from './config'

const debug = Debug('api-docs-gen:generator')

/**
 * generate markdown contents
 *
 * @param input input paths
 * @param output output full path
 * @param style generate style, see {@link GenerateStyle}
 * @param config configration, see {@link Config}
 */
export async function generate(
  input: string[],
  output: string,
  style: GenerateStyle,
  config: Config,
  callback?: (pkgname: string, filename: string) => void
): Promise<void> {
  debug(`config`, config)
  debug('style', style)

  const apiModel = new ApiModel()
  for (const target of input) {
    debug(`generate from ${target} ...`)
    const apiPackage = loadPackage(target, apiModel)

    let result = config.processor(
      apiModel,
      apiPackage,
      style,
      config.linkReferencer! // eslint-disable-line @typescript-eslint/no-non-null-assertion
    )

    if (isString(result)) {
      result = [
        {
          filename: 'api.md',
          body: result
        }
      ]
    }

    if (!Array.isArray(result)) {
      throw new Error('Not supported processor result type')
    }

    let outputTarget = output
    if (style === 'directory') {
      outputTarget = path.resolve(output, apiPackage.displayName)
      try {
        await mkdir(outputTarget)
      } catch (e) {
        throw new Error(`Cannot make '${outputTarget} directory: ${e.message}`)
      }
    }

    for (const { filename, body } of result) {
      const filepath = path.resolve(
        outputTarget,
        style === 'prefix' ? `${apiPackage.displayName}-${filename}` : filename
      )
      await writeFile(filepath, body)
      callback && callback(apiPackage.displayName, filepath)
    }
  }
}
