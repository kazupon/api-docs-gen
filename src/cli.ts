import path from 'path'
import meow from 'meow'
import chalk from 'chalk'
import { debug as Debug } from 'debug'
import { generate } from './generator'

const debug = Debug('api-docs-gen:cli')

export const flags = {
  output: {
    type: 'string',
    alias: 'o',
    isRequired: false
  },
  config: {
    type: 'string',
    alias: 'c',
    isRequired: false
  }
} as const

const cli = meow(
  `
  Usage
    $ api-docs-gen <input>

  Options
    --config, -c configuration file
    --output, -o output dierectory that is markdown contents

  Examples
    $ api-docs-gen my-library-api.json
    $ api-docs-gen my-library-api.json --output ./docs
    $ api-docs-gen my-library-api.json --config docsgen.config.js
`,
  {
    flags
  }
)

// check input
const input = cli.input
if (!input.length) {
  console.error(chalk.red(`[api-docs-gen] not specified docs model`))
  process.exit(1)
}

// resolve output
const output =
  cli.flags.output != null ? path.resolve(cli.flags.output) : process.cwd()
debug(`output`, output)

// config
const config = cli.flags.config
debug(`config`, config)

// run
try {
  ;(async () => {
    await generate(input, output, config)
  })()
} catch (e) {
  console.error(chalk.red(`[api-docs-gen] ${e.message}`))
  process.exit(1)
}

process.on('uncaughtException', err => {
  console.error(chalk.red(`[api-docs-gen] Uncaught exception: ${err}\n`))
  process.exit(1)
})

process.on('unhandledRejection', (reason, p) => {
  console.error(
    chalk.red(`[api-docs-gen] Unhandled rejection at: ${p}, reason: ${reason}`)
  )
  process.exit(1)
})
