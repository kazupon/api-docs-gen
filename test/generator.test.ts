import path from 'path'
import { ApiModel } from '@microsoft/api-extractor-model'

/**
 * setup ApiPackage Stub data
 */
const API_JSON1 = path.resolve(__dirname, './fixtures/library1.api.json')
const API_JSON2 = path.resolve(__dirname, './fixtures/utilities.api.json')
const model = new ApiModel()
const API_MODEL_FILES = [API_JSON1, API_JSON2]
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const PACKAGES: any = API_MODEL_FILES.reduce((packages, file) => {
  return Object.assign(packages, { [file]: model.loadPackage(file) })
}, {})

// mock 'resolver' module
jest.mock('../src/resolver/index', () => {
  const { multi } = jest.requireActual('../src/resolver/index')
  return {
    multi,
    loadPackage: jest.fn().mockImplementation(modelPath => PACKAGES[modelPath]) // stub
  }
})

// mock 'utils' modules
jest.mock('../src/utils', () => ({
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ...jest.requireActual<any>('../src/utils'),
  mkdir: jest.fn(),
  writeFile: jest.fn()
}))
import { mkdir, writeFile } from '../src/utils'

import { DefaultConfig, GenerateStyle } from '../src/config'
import { generate } from '../src/generator'

afterEach(() => {
  jest.clearAllMocks()
})

test('generate prefix style contents', async () => {
  const input = [API_JSON1, API_JSON2]
  const output = path.resolve(__dirname, './')
  await generate(input, output, {
    style: GenerateStyle.Prefix,
    config: DefaultConfig
  })

  expect(mkdir).toHaveBeenCalledTimes(7)
  expect(writeFile).toHaveBeenCalledTimes(7)
  // @ts-ignore TS2339: Property 'mock' does not exist on type '(path: string, data: string) => Promise<void>'.
  for (const arg of writeFile.mock.calls) {
    expect(
      [
        path.resolve(output, './library1-enum.md'),
        path.resolve(output, './library1-function.md'),
        path.resolve(output, './library1-variable.md'),
        path.resolve(output, './library1-class.md'),
        path.resolve(output, './library1-interface.md'),
        path.resolve(output, './library1-typealias.md'),
        path.resolve(output, './utilities-function.md')
      ].includes(arg[0])
    ).toBe(true)
    expect(arg[1]).toMatchSnapshot()
  }
})

test('generate directory style contents', async () => {
  const input = [API_JSON1, API_JSON2]
  const output = path.resolve(__dirname, './')
  await generate(input, output, {
    style: GenerateStyle.Directory,
    config: DefaultConfig
  })

  expect(mkdir).toHaveBeenCalledTimes(7)

  const passedArgs = [
    path.resolve(output, './library1/enum.md'),
    path.resolve(output, './library1/function.md'),
    path.resolve(output, './library1/variable.md'),
    path.resolve(output, './library1/class.md'),
    path.resolve(output, './library1/interface.md'),
    path.resolve(output, './library1/typealias.md'),
    path.resolve(output, './utilities/function.md')
  ]
  expect(writeFile).toHaveBeenCalledTimes(7)
  // @ts-ignore TS2339: Property 'mock' does not exist on type '(path: string, data: string) => Promise<void>'.
  for (const arg of writeFile.mock.calls) {
    expect(passedArgs.includes(arg[0])).toBe(true)
    expect(arg[1]).toMatchSnapshot()
  }
})
