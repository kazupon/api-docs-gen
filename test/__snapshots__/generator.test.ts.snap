// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate directory style contents 1`] = `
"# TypeAlias

## A

A

**Signature:**
\`\`\`typescript
export declare type A = string;
\`\`\`

## B

[A](./typealias#a)

**Signature:**
\`\`\`typescript
export declare type B = number;
\`\`\`

## FallbackLocale

Falback Locale

**Signature:**
\`\`\`typescript
export declare type FallbackLocale = Locale | Locale[] | {
    [locale in string]: Locale[];
} | false;
\`\`\`

### Remarks

This is remarks of \`Fallback Locale\`

## Locale

Locale

**Signature:**
\`\`\`typescript
export declare type Locale = string;
\`\`\`

## LocaleMessage

Locale Message resources

**Signature:**
\`\`\`typescript
export declare type LocaleMessage = string | LocaleMessage[];
\`\`\`

"
`;

exports[`generate directory style contents 2`] = `
"# Function

## add

add function : \`x\`

**Signature:**
\`\`\`typescript
export declare function add(a: number, b: number): number;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target \`1\` |
| b | number | target 2 |

### Returns

 result as \`a\` + \`b\`

### Throws

- \`SyntaxError\` this is syntax error
- \`Error\` this is general error

### Remarks

This is add function remarks: [ErrorCodes](./enum#errorcodes). See [add](https://foo.bar.com/api/add).

### Examples

example of \`add\` function:
\`\`\`javascript
console.log(add(1, 1))
\`\`\`


"
`;

exports[`generate directory style contents 3`] = `
"# Interface

## Calculatable

Calculatable interface

**Signature:**
\`\`\`typescript
export interface Calculatable 
\`\`\`


### Methods

#### add

add method

**Signature:**
\`\`\`typescript
add(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |


### Properties

#### PI

PI

**Signature:**
\`\`\`typescript
PI: number;
\`\`\`


"
`;

exports[`generate directory style contents 4`] = `
"# Class

## Calculator

Calculator class

**Signature:**
\`\`\`typescript
export declare class Calculator implements Calculatable 
\`\`\`

### Remarks

This is remarks of Calculator class

### Examples


\`\`\`javascript
const c = new Calculator()
const v1 = c.add(1, 1)
const v2 = c.sub(1, 1)
\`\`\`



### Constructor

Conssutructor of usage

**Signature:**
\`\`\`typescript
constructor(type: string);
\`\`\`


### Methods

#### add

add method

**Signature:**
\`\`\`typescript
add(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |

#### sub

sub method

**Signature:**
\`\`\`typescript
sub(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |


### Properties

#### PI

PI

**Signature:**
\`\`\`typescript
PI: number;
\`\`\`

#### type

calcurator types 'simple'

**Signature:**
\`\`\`typescript
type: string;
\`\`\`


"
`;

exports[`generate directory style contents 5`] = `
"# Variable

## Config

Configrations 

**Signature:**
\`\`\`typescript
Config: {
    [name: string]: unknown;
}
\`\`\`

## VERSION

app version

**Signature:**
\`\`\`typescript
VERSION = \\"1.0.0\\"
\`\`\`

### Remarks

you can lookup application version that is semver format.

### Examples


\`\`\`javascript
console.log(VERSION)
\`\`\`


"
`;

exports[`generate directory style contents 6`] = `
"# Enum

## ErrorCodes

Error Code

**Signature:**
\`\`\`typescript
export declare enum ErrorCodes 
\`\`\`

### Members

| Member | Value| Description |
| --- | --- | --- |
| InvalidFormat | 1 | Invalid format |
| Succcess | 0 | Success |

## TokenChars

Token Characters

**Signature:**
\`\`\`typescript
export declare enum TokenChars 
\`\`\`

### Members

| Member | Value| Description |
| --- | --- | --- |
| Modulo | \\"%\\" | Modulo charactor |
| Plus | \\"+\\" | Plus charactor |

### Remarks

This is remarks of Token Chararaceters

"
`;

exports[`generate directory style contents 7`] = `
"# Function

## isBoolean

Whether the value is Boolean or not

**Signature:**
\`\`\`typescript
export declare function isBoolean(val: unknown): val is boolean;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isFunction

Whether the value is Function or not

**Signature:**
\`\`\`typescript
export declare function isFunction(val: unknown): val is Function;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isObject

Whether the value is Object or not

**Signature:**
\`\`\`typescript
export declare function isObject(val: unknown): val is Record<string, unknown>;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isPlainObject

Whether the value is Plain Object or not

**Signature:**
\`\`\`typescript
export declare function isPlainObject(val: unknown): val is Record<string, unknown>;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isPromise

Whether the value is Promise or not

**Signature:**
\`\`\`typescript
export declare function isPromise<T = any>(val: unknown): val is Promise<T>;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isString

Whether the value is String or not

**Signature:**
\`\`\`typescript
export declare function isString(val: unknown): val is string;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isSymbol

Whether the value is Symbol or not

**Signature:**
\`\`\`typescript
export declare function isSymbol(val: unknown): val is symbol;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

"
`;

exports[`generate prefix style contents 1`] = `
"# TypeAlias

## A

A

**Signature:**
\`\`\`typescript
export declare type A = string;
\`\`\`

## B

[A](./library1-typealias#a)

**Signature:**
\`\`\`typescript
export declare type B = number;
\`\`\`

## FallbackLocale

Falback Locale

**Signature:**
\`\`\`typescript
export declare type FallbackLocale = Locale | Locale[] | {
    [locale in string]: Locale[];
} | false;
\`\`\`

### Remarks

This is remarks of \`Fallback Locale\`

## Locale

Locale

**Signature:**
\`\`\`typescript
export declare type Locale = string;
\`\`\`

## LocaleMessage

Locale Message resources

**Signature:**
\`\`\`typescript
export declare type LocaleMessage = string | LocaleMessage[];
\`\`\`

"
`;

exports[`generate prefix style contents 2`] = `
"# Function

## add

add function : \`x\`

**Signature:**
\`\`\`typescript
export declare function add(a: number, b: number): number;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target \`1\` |
| b | number | target 2 |

### Returns

 result as \`a\` + \`b\`

### Throws

- \`SyntaxError\` this is syntax error
- \`Error\` this is general error

### Remarks

This is add function remarks: [ErrorCodes](./library1-enum#errorcodes). See [add](https://foo.bar.com/api/add).

### Examples

example of \`add\` function:
\`\`\`javascript
console.log(add(1, 1))
\`\`\`


"
`;

exports[`generate prefix style contents 3`] = `
"# Interface

## Calculatable

Calculatable interface

**Signature:**
\`\`\`typescript
export interface Calculatable 
\`\`\`


### Methods

#### add

add method

**Signature:**
\`\`\`typescript
add(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |


### Properties

#### PI

PI

**Signature:**
\`\`\`typescript
PI: number;
\`\`\`


"
`;

exports[`generate prefix style contents 4`] = `
"# Class

## Calculator

Calculator class

**Signature:**
\`\`\`typescript
export declare class Calculator implements Calculatable 
\`\`\`

### Remarks

This is remarks of Calculator class

### Examples


\`\`\`javascript
const c = new Calculator()
const v1 = c.add(1, 1)
const v2 = c.sub(1, 1)
\`\`\`



### Constructor

Conssutructor of usage

**Signature:**
\`\`\`typescript
constructor(type: string);
\`\`\`


### Methods

#### add

add method

**Signature:**
\`\`\`typescript
add(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |

#### sub

sub method

**Signature:**
\`\`\`typescript
sub(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 |
| b | number | target 2 |


### Properties

#### PI

PI

**Signature:**
\`\`\`typescript
PI: number;
\`\`\`

#### type

calcurator types 'simple'

**Signature:**
\`\`\`typescript
type: string;
\`\`\`


"
`;

exports[`generate prefix style contents 5`] = `
"# Variable

## Config

Configrations 

**Signature:**
\`\`\`typescript
Config: {
    [name: string]: unknown;
}
\`\`\`

## VERSION

app version

**Signature:**
\`\`\`typescript
VERSION = \\"1.0.0\\"
\`\`\`

### Remarks

you can lookup application version that is semver format.

### Examples


\`\`\`javascript
console.log(VERSION)
\`\`\`


"
`;

exports[`generate prefix style contents 6`] = `
"# Enum

## ErrorCodes

Error Code

**Signature:**
\`\`\`typescript
export declare enum ErrorCodes 
\`\`\`

### Members

| Member | Value| Description |
| --- | --- | --- |
| InvalidFormat | 1 | Invalid format |
| Succcess | 0 | Success |

## TokenChars

Token Characters

**Signature:**
\`\`\`typescript
export declare enum TokenChars 
\`\`\`

### Members

| Member | Value| Description |
| --- | --- | --- |
| Modulo | \\"%\\" | Modulo charactor |
| Plus | \\"+\\" | Plus charactor |

### Remarks

This is remarks of Token Chararaceters

"
`;

exports[`generate prefix style contents 7`] = `
"# Function

## isBoolean

Whether the value is Boolean or not

**Signature:**
\`\`\`typescript
export declare function isBoolean(val: unknown): val is boolean;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isFunction

Whether the value is Function or not

**Signature:**
\`\`\`typescript
export declare function isFunction(val: unknown): val is Function;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isObject

Whether the value is Object or not

**Signature:**
\`\`\`typescript
export declare function isObject(val: unknown): val is Record<string, unknown>;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isPlainObject

Whether the value is Plain Object or not

**Signature:**
\`\`\`typescript
export declare function isPlainObject(val: unknown): val is Record<string, unknown>;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isPromise

Whether the value is Promise or not

**Signature:**
\`\`\`typescript
export declare function isPromise<T = any>(val: unknown): val is Promise<T>;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isString

Whether the value is String or not

**Signature:**
\`\`\`typescript
export declare function isString(val: unknown): val is string;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

## isSymbol

Whether the value is Symbol or not

**Signature:**
\`\`\`typescript
export declare function isSymbol(val: unknown): val is symbol;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| val | unknown | target value |

### Returns

 boolean value

"
`;
